File: express/src/lib/sentry.ts
import * as Sentry from '@sentry/node';
import type { ErrorRequestHandler, RequestHandler } from 'express';

export const SENTRY_PURPOSE = 'sentry';

const DEFAULT_ENVIRONMENT = 'production';

export interface SentryRuntimeConfig {
  dsn: string;
  environment: string;
  release: string;
}

export function normalizeSentryConfig(
  raw: Record<string, unknown> | null,
): SentryRuntimeConfig {
  const dsn = typeof raw?.dsn === 'string' ? raw.dsn : '';
  const environment =
    typeof raw?.environment === 'string' && raw.environment
      ? raw.environment
      : DEFAULT_ENVIRONMENT;
  const release = typeof raw?.release === 'string' ? raw.release : '';
  return { dsn, environment, release };
}

let initialized = false;

export function initSentry(runtime: SentryRuntimeConfig): boolean {
  if (initialized) return true;
  if (!runtime.dsn) return false;

  Sentry.init({
    dsn: runtime.dsn,
    environment: runtime.environment,
    release: runtime.release || undefined,
    serverName: 'backend',
    tracesSampleRate: 0,
    profilesSampleRate: 0,
    sendDefaultPii: false,
    integrations: [
      Sentry.httpIntegration(),
      Sentry.onUncaughtExceptionIntegration({
        exitEvenIfOtherHandlersAreRegistered: false,
      }),
      Sentry.onUnhandledRejectionIntegration({ mode: 'warn' }),
    ],
    beforeSend(event, hint) {
      const err = hint?.originalException as
        | { name?: string; statusCode?: number }
        | undefined;
      if (err?.name === 'NotFoundError' || err?.name === 'BusinessRuleError')
        return null;
      const status = err?.statusCode;
      if (typeof status === 'number' && status >= 400 && status < 500)
        return null;
      return event;
    },
    beforeSendTransaction(event) {
      const path = event.transaction;
      if (typeof path === 'string' && path.startsWith('/api/health'))
        return null;
      return event;
    },
  });

  initialized = true;
  return true;
}

export const sentryRequestHandler: RequestHandler = (req, _res, next) => {
  if (!initialized) {
    next();
    return;
  }
  Sentry.getCurrentScope().setTag(
    'request_id',
    (req as { id?: string }).id ?? '',
  );
  next();
};

export const sentryErrorHandler: ErrorRequestHandler = (
  err,
  _req,
  _res,
  next,
) => {
  if (!initialized) {
    next(err);
    return;
  }
  const status = (err as { statusCode?: number }).statusCode;
  if (typeof status !== 'number' || status >= 500) {
    Sentry.captureException(err);
  }
  next(err);
};

export const __sentryInternals = {
  reset(): void {
    initialized = false;
  },
};

export const SentryNode = Sentry;

File: fastapi/src/_sentry.py
from typing import Any

import sentry_sdk
from sentry_sdk.integrations.asyncio import AsyncioIntegration
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.httpx import HttpxIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration

SENTRY_PURPOSE = "sentry"

_DEFAULT_ENVIRONMENT = "production"

_DROP_EXCEPTION_NAMES = {
    "BusinessRuleError",
    "NotFoundError",
    "RequestValidationError",
    "HTTPException",
    "AuthenticationError",
}

_DROP_TRANSACTION_PREFIXES = ("GET /api/health",)


def _before_send(event: Any, hint: dict[str, Any]) -> Any:
    exc_info = hint.get("exc_info") if hint else None
    if exc_info:
        exc = exc_info[1]
        if exc is not None:
            name = type(exc).__name__
            if name in _DROP_EXCEPTION_NAMES:
                return None
            status_code = getattr(exc, "status_code", None)
            if isinstance(status_code, int) and 400 <= status_code < 500:
                return None
    return event


def _before_send_transaction(event: Any, _hint: dict[str, Any]) -> Any:
    transaction = event.get("transaction")
    if isinstance(transaction, str) and transaction.startswith(_DROP_TRANSACTION_PREFIXES):
        return None
    return event


def init_sentry(config: dict[str, Any] | None) -> bool:
    dsn = str((config or {}).get("dsn", "")).strip()
    if not dsn:
        return False
    environment = str((config or {}).get("environment") or _DEFAULT_ENVIRONMENT)
    release = str((config or {}).get("release") or "") or None
    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,
        server_name="backend",
        traces_sample_rate=0.0,
        send_default_pii=False,
        max_request_body_size="small",
        before_send=_before_send,
        before_send_transaction=_before_send_transaction,
        integrations=[
            FastApiIntegration(transaction_style="endpoint"),
            StarletteIntegration(transaction_style="endpoint"),
            AsyncioIntegration(),
            HttpxIntegration(),
            SqlalchemyIntegration(),
        ],
    )
    return True
